service/frontend/nexus_incoming_service_registry.go

```diff
@@ -0,0 +1,406 @@
+			return nil, serviceerror.NewFailedPrecondition("could not find service indicated by nexus list incoming services next page token")
+		}
+	}
+
+	endIdx := startIdx + int(request.PageSize)
+	if endIdx > len(r.serviceIDsSorted) {
+		endIdx = len(r.serviceIDsSorted)
+	}
+
+	var nextPageToken []byte
+	if endIdx < len(r.services) {
+		nextPageToken = []byte(r.serviceIDsSorted[endIdx])
+	}
+
+	return &operatorservice.ListNexusIncomingServicesResponse{
+		NextPageToken: nextPageToken,
+		Services:      r.services[startIdx:endIdx],
+	}, nil
+}
+
+func (r *registry) isTableVersionCurrent(ctx context.Context) bool {
+	actualTableVersion, err := r.persistence.GetNexusIncomingServicesTableVersion(ctx)
+	if err != nil {
+		return false
+	}
+
+	r.RLock()
+	defer r.RUnlock()
+	return r.tableVersion == actualTableVersion
+}
+
+func (r *registry) fetchServices(ctx context.Context) error {
+	var currentPageToken []byte
+	pagingWasReset := false // used to ensure write lock is only acquired once
+
+	for ctx.Err() == nil {
+		r.RLock()
+		requestTableVersion := r.tableVersion
+		r.RUnlock()
+
+		shouldWait := currentPageToken == nil && requestTableVersion != 0
+
+		resp, err := r.matchingClient.ListNexusIncomingServices(ctx, &matchingservice.ListNexusIncomingServicesRequest{
+			LastKnownTableVersion: requestTableVersion,
+			NextPageToken:         currentPageToken,
+			PageSize:              loadServicesPageSize,
+			Wait:                  shouldWait,
+		})
+		if err != nil {
+			if errors.Is(err, p.ErrNexusTableVersionConflict) {
+				// indicates table was updated during paging, so reset and start from the beginning
+				currentPageToken = nil
+				pagingWasReset = true
+				r.invalidateCachedServicesLocked()
+				continue
+			}
+			// for all other errors, fall back to reading directly from persistence
+			if currentPageToken == nil && !pagingWasReset {
+				r.Lock()
+			}
+			persistenceErr := r.loadServicesLocked(ctx)
+			if persistenceErr == nil {
+				r.setServiceDataReady(persistenceErr)
+			}
+			r.Unlock()
+			return persistenceErr
+		}
+
+		if shouldWait && requestTableVersion == resp.TableVersion {
+			// long poll returned with no changes
+			return nil
+		}
+
+		if currentPageToken == nil && !pagingWasReset {
+			// acquire lock once upon getting first page and release once all pagination is complete
+			r.Lock()
+		}
+
+		if shouldWait {
+			// long poll returned new data, starting with first page
+			r.invalidateCachedServicesLocked()
+		}
+
+		currentPageToken = resp.NextPageToken
+		r.tableVersion = resp.TableVersion
+		r.updateCachedServicesLocked(resp.Entries)
+
+		if resp.NextPageToken == nil {
+			break
+		}
+	}
+
+	r.setServiceDataReady(ctx.Err())
+	r.Unlock()
+	return ctx.Err()
+}
+
+// Loads all incoming services into memory from persistence.
+// Used as a fallback if matching service is unavailable.
+// Always replaces entire in-memory view by paging from the start.
+func (r *registry) loadServicesLocked(ctx context.Context) error {
```

Comment:
This function name should mention persistence because it's not immediately clear without reading the docstring or implementation.